[AoTI Minifier] UX Improvement

yushangdi opened the pull request:
Summary:
- When a user specify `TORCHINDUCTOR_MAX_AUTOTUNE=1` env variable, we add `config.max_autotune=True` to the generated minifier_launcher
- We should do this to other inductor configs as well in a followup Diff

Currently in dynamo and aoti minifier, if a config is overwritten by an env variable, the config will not show up in the config list in the minifier_launcher.py file. As a result, when running the minifier_launcher, they need to re-apply the same env variable.
 This is:
1) not convenient for the users
2) if they copy-paste the minifier_launcher.py to us without including the env variable, we could be confused and not able to reproduce the error.

Underlying implementation change:

- Add `env_default` parameter to `codegen_config()`. If set, configs overriden by the env are not considered default.

Test Plan:
```
 buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:utils -- -r test_codegen_config
```

Differential Revision: D67299312


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

desertfire commented on torch/utils/_config_module.py:
```diff
@@ -432,19 +454,20 @@ def _get_dict(
             skip_default does two things. One if a key has not been modified
                 it skips it. The other is it modified the logging behaviour
                 to match what codegen already did for modified skipped keys
+            env_default: If set, configs overriden by the env are not considered default.
```
I think this should always be True?

yushangdi replied:
Currently, the test is written such that these configs are skipped by skip_default. If we change this to always be True, we need to change the existing tests. I'm checking with test writer (Colin) if we can change it.

yushangdi replied:
I'll change it to always be True

desertfire commented:
Adding more inductor reviewers

eellison commented on torch/_inductor/config.py:
```diff
@@ -293,7 +293,7 @@ def bundle_triton_into_fx_graph_cache_default() -> Optional[bool]:
 
 # enable slow autotuning passes to select algorithms
-max_autotune = os.environ.get("TORCHINDUCTOR_MAX_AUTOTUNE") == "1"
+max_autotune = Config(False, env_name_default="TORCHINDUCTOR_MAX_AUTOTUNE")
```
Why do we need this ? the existing minifier serializes config settings without this change

yushangdi replied:
@eellison 

The problem is that the minifier cannot produce the same config as the run time config.

For example, if you run the script below like

If you just run the code below directly, like `TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py`, the configs that the minifier dumps doesn't have `torch._inductor.config.max_autotune = True`, even though you have env variable `TORCHINDUCTOR_MAX_AUTOTUNE=1`.


This is because the config module takes the ` os.environ.get("TORCHINDUCTOR_MAX_AUTOTUNE") == "1"` as default and skipped this config, which is wrong. The true default value is False, and the user actually changed the value of this config with an env variable. 

```
import torch
from torch._inductor import config as inductor_config
class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = torch.nn.Linear(10, 16)
        self.relu = torch.nn.ReLU()
        self.sigmoid = torch.nn.Sigmoid()
    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.sigmoid(x)
        return x
inductor_config.aot_inductor.dump_aoti_minifier = True
torch._inductor.config.triton.inject_relu_bug_TESTING_ONLY = "compile_error"
with torch.no_grad():
    model = Model().to("cuda")
    example_inputs = (torch.randn(8, 10).to("cuda"),)
```

yushangdi replied:
Most of the changes here is to add a flag when dumping the configs, so we don't skip the configs that are changed by the setting the env variables.

eellison replied:
I think we should be able to do the above without the `max_autotune = Config(False, env_name_default="TORCHINDUCTOR_MAX_AUTOTUNE")` changes

yushangdi replied:
> I think we should be able to do the above without the `max_autotune = Config(False, env_name_default="TORCHINDUCTOR_MAX_AUTOTUNE")` changes.. But I guess this is the standard way, so fine with me..

@eellison yeah I agree, we could do it another way, but since there is the `Config` class already, I thought we might as well just use it.

eellison replied:
It does work, but I would still prefer we do the general thing that covers all configs instead of just this specific one.

eellison reviewed (commented):
Can you reuse the existing config serialization from minifier? i would not expect this many changes

jansel commented on torch/_dynamo/config.py:
```diff
@@ -421,6 +421,14 @@ def default_debug_dir_root():
     if DEBUG_DIR_VAR_NAME in os.environ:
         return os.path.join(os.environ[DEBUG_DIR_VAR_NAME], "torch_compile_debug")
     elif is_fbcode():
+        fbcode_dir = os.path.join(os.path.expanduser("~"), "fbsource", "fbcode")
```
Is this path general?  Should this be cwd or home?  Putting it in the fbsource tree seems a bit weird.

yushangdi replied:
It's intended to be in the fbcode. The generated TARGETS file and python files need to be in fbsource for them to be compiled by buck.

yushangdi replied:
I guess I could also make it cwd.

jansel commented on torch/_inductor/config.py:
```diff
@@ -1380,3 +1380,15 @@ class test_configs:
 # adds patch, save_config, etc
 install_config_module(sys.modules[__name__])
+
+# def _set_config_from_env(config, env) -> None:
+#     """
+#     Set config to True if env is set to 1.
+#     We use the if statement to avoid calling setattr if env is not set.
+#     This helps config.codegen_config() to only generate configs that are not
+#     the default value.
+#     """
+#     if os.environ.get(env) == "1":
+#         setattr(sys.modules[__name__], config, True)
+
+# _set_config_from_env("max_autotune", "TORCHINDUCTOR_MAX_AUTOTUNE")
```
Dead code?

jansel commented on torch/utils/_config_module.py:
```diff
@@ -407,14 +407,36 @@ def _set_alias_val(self, entry: _ConfigEntry, val: Any) -> None:
         ignored_keys: Optional[List[str]] = None,
         ignored_prefixes: Optional[List[str]] = None,
         skip_default: bool = False,
+        env_default: bool = False,
```
Does anything break if we just turn this on always?  If so let's just remove the option and make it default.

yushangdi replied:
@c00w Do you know if we can make this `env_default` default to True? (If a config's env value is different from default value, it cannot be skipped by `skip_default`).

Currently, the test is written such that these configs are skipped by `skip_default`.

yushangdi replied:
I think it's probably ok. I'll make it default then.

jansel commented on torch/_dynamo/repro/aoti.py:
```diff
@@ -455,7 +455,7 @@ def common_flags(parser):
 
     subparsers = parser.add_subparsers(
-        dest="command", metavar="{run,minify,analyze}", required=True
+        dest="command", metavar="{run,minify}", required=True
```
Was this unused?

yushangdi replied:
yeah. the AoTI minifier doesn't have an analyze mode yet.

yushangdi replied:
yeah, AoTI minifier doesn't have an analyze mode yet

yushangdi commented:
@jansel @desertfire Pin for review on the updated version of the PR. Thank you!

yushangdi committed:
[AoTI Minifier] UX Improvement (#143330)

Summary:

- When a user specify env variables that include "TORCH", "INDUCTOR", "DYNAMO" or "TRITON", we add  those env variable definitions to the generated minifier_launcher


Example env variables generated:

```
import os
os.environ['TORCHINDUCTOR_MAX_AUTOTUNE'] = '1'
os.environ['PYTORCH_DDP_USE_SIDE_STREAM'] = '0'
os.environ['TRITON_CACHE_MANAGER'] = 'triton.runtime.cache:RemoteCacheManager'
os.environ['TORCHINDUCTOR_CACHE_DIR'] = '/tmp/torchinductor_shangdiy'
os.environ['TRITON_REMOTE_CACHE_BACKEND'] = 'triton.fb.fb_memcache:FbMemcacheRemoteKernelCache'
```

We skip some env vars like below, because they are generated per compile.

```

os.environ['TRITON_LIBDEVICE_PATH'] = '/data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/585c17c94f036562/tutorials/pytorch/nntest/__torchtest__/torchtest#link-tree/triton/fb/libdevice.10.bc'
os.environ['TRITON_PTXAS_PATH'] = '/data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/585c17c94f036562/tutorials/pytorch/nntest/__torchtest__/torchtest#link-tree/triton/fb/ptxas'
os.environ['TRITON_LIBCUDA_PATH'] = '/usr/local/fbcode/platform010/lib'
```

Currently in dynamo and aoti minifier, if a config is overwritten by an env variable, the config will not show up in the config list in the minifier_launcher.py file. As a result, when running the minifier_launcher, they need to re-apply the same env variable.
 This is:
1) not convenient for the users
2) if they copy-paste the minifier_launcher.py to us without including the env variable, we could be confused and not able to reproduce the error.

We also make the following change in `codegen_config()`:

- Add `env_default` parameter to `codegen_config()`. If set, configs overriden by the env are not considered default.

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:utils -- -r test_codegen_config
```

Reviewed By: eellison

Differential Revision: D67299312

eellison reviewed (approved):
🚀 🚀 🚀

yushangdi commented:
> This pull request has been **reverted** by [66ce13b](https://github.com/pytorch/pytorch/commit/66ce13b4974ad7b470f60bf871604bff1c87f4ec). To re-land this change, please open another pull request, assignthe same reviewers, fix the CI failures that caused the revert and make sure that the failing CI runs on the PR by applying the proper ciflow label (e.g., ciflow/trunk).

Note: this Diff is only partially reverted in [66ce13b]. The majority of the Diff is not reverted.
